Patcher/UnityInstallation.cs

```diff
@@ -0,0 +1,98 @@
+namespace Patcher
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Text.RegularExpressions;
+
+    /// <summary>
+    /// Represents an Unity Installation on the Disk. 
+    /// </summary>
+    public class UnityInstallation
+    {
+        private readonly string _installationLocation;
+
+        private readonly OperatingSystem _operatingSystem;
+
+        /// <summary>
+        /// Creates a new Unity installation
+        /// </summary>
+        /// <param name="installationLocation">The Path to this specific installation on the File System</param>
+        /// <param name="operatingSystem"></param>
+        public UnityInstallation(string installationLocation, OperatingSystem operatingSystem)
+        {
+            _installationLocation = installationLocation;
+            _operatingSystem = operatingSystem;
+        }
+
+        /// <summary>
+        /// The Version of the Unity Installation
+        /// </summary>
+        public string Version => Path.GetFileName(_installationLocation);
+
+        /// <summary>
+        /// Gets the Path to the Unity executable on the disk.
+        /// </summary>
+        /// <returns></returns>
+        public string ExecutablePath() => Path.Combine(_installationLocation, _operatingSystem switch
+        {
+            OperatingSystem.Windows => @"Editor\Unity.exe",
+            OperatingSystem.MacOS => "Unity.app/Contents/MacOS/Unity",
+            OperatingSystem.Linux => "Unity",
+            _ => throw new ArgumentOutOfRangeException(nameof(_operatingSystem))
+        });
+
+        public bool IsSupported(IEnumerable<PatchInfo> patches)
+        {
+            return GetPatch(patches) != null;
+        }
+
+        public PatchInfo GetPatch(IEnumerable<PatchInfo> patches)
+        {
+            foreach (PatchInfo patch in patches)
+            {
+                if (Regex.IsMatch(this.Version, patch.Version))
```

Comment:
`this` isn't required here. Also, assuming this gets passed values such as `2020.1`, does this work? I didn't realize you could use regular expressions like this.

Reply:
Overlooked the `this` when I was trying to remove all of them.

The `Regex.IsMatch` returns `true` if the pattern (patch.Version) is contained at least once in the input (Version), examples would be:
| patch.Version | Version | Result |
| ------------- | ------------ | ------ |
| 2018.4            | 2018.4.1f1    | `true`  |
| 2018                | 2018.4.22f1 | `true`  |
| 2020.2            | 2020.2.0a2 | `true`  |
| 2019.3.6f1     | 2019.3.6f1   | `true`  |
| 2019.3.6f1     | 2019.3.7f1   | `false` |

This allows us to specify patches for more than once specific Version of Unity, which is handy because I saw that for Windows the entire 2018 Release Cycle has the same PatchInfo.

One thing that might come back to bite us is that it will return the first patch.Version that matches, which might lead to problems when for Example 2020.2.1f1 needs a special Patch but the rest of the 2020.2 Cycle has the same Patch, then it might return the general 2020.2 one instead of the 2020.2.1f1 one. But I have not seen something like that thus far, so I figured we can work that out when it comes to it.

Reply:
It might also make sense to add some Unit Tests for that Method, if you tell me which Framework you prefer for Testing I can add some if you want.